### ISSUE TITLE
Provide mechanisms for users to wire in custom implementations of TestRunner and Suite Runner

### ISSUE THREAD

**Issue by krmahadevan** [testng-team/testng#2632]
Currently TestNG does not provide any means for an end user to inject the below objects

* `org.testng.SuiteRunner` which represents a `<suite>` that is going to be run (This is different from the `XmlSuite` which only represents the contents of the suite but not the runner that is going to run the suite). It is currently being instantiated via `org.testng.TestNG#createSuiteRunner`
* `TestRunner` which represents a `<test>` that is going to be run.

There is more details in this Google groups thread : https://groups.google.com/g/testng-users/c/CFglgMdqxww/m/u-BZELoJFgAJ

@juherr, @cbeust - What do you think about this ? 

The use case from the user seems to be genuine because TestNG is a test runner. It should not have any opinion on what sort of tests a user runs. Also in the real world there could be a lot of practical applications for being able to use TestNG in a performance testing scenario because:

1. In the Java world there aren't too many developer friendly performance tools. We have JMeter, but its xml driven and is a pain when it comes to adding simple programming constructs. And then there is Gatling which is scala based which means I now have an additional learning curve.
2.  For perf tests, I should be able to leverage my functional tests itself (or atleast some trimmed down variants of it) instead of reinventing the wheel and defining additional tests for them.

**Comment by juherr** [comment]
I'm not sure to understand the need.

If someone wants to reuse internal classes it will be its responsibility but I'm not sure it will be a good idea to add features to the current API.
On the opposite, if some aspects of the internal code have values we can debate on a clear new API and expose it under its own module or library.

What could be the minimal scope here and what could be its dependencies?

**Comment by krmahadevan** [comment]
@juherr 

> If someone wants to reuse internal classes it will be its responsibility but I'm not sure it will be a good idea to add features to the current API.

Ok. Let me rephrase. We would need to abstract out a TestRunner capabilities and a SuiteRunner capabilities as perhaps an interface and then provide a way wherein we can accept instantiation of both of these objects from the user (optionally). I am not asking that we expose the implementation, but just saying that we should have a way wherein a user can provide their own version of `TestRunner` and `SuiteRunner` which we can use (similar to what we do with listeners).

The internals of TestNG stay internal. But if a user wants to plug in their customised `TestRunner` or `SuiteRunner` then he/she should be able to do that.

**Comment by juherr** [comment]
IMO, `*Runner` is part of the internal which is not yet exposed on a public API. 
I'm not sure adding a new API is a good idea because it will make it more difficult to change it later if needed.

Instead, one could create a new test engine based on some internal TestNG code and we can help by providing needed classes in a dedicated library.

**Comment by krmahadevan** [comment]
@juherr - The TestRunner is actually speaking an implementation of a bunch of interfaces (most of them are already exposed)

```
public class TestRunner
    implements ITestContext, ITestResultNotifier, IThreadWorkerFactory<ITestNGMethod>,
    IConfigEavesdropper {}

```
public class SuiteRunner implements ISuite, IInvokedMethodListener {}
```

I was thinking that we can perhaps expose 1 single interface which would extend all the other interfaces and call that as 

* TestRunner
* SuiteRunner

Something like below

```
public interface TestRunner extends ITestContext, ITestResultNotifier, IThreadWorkerFactory<ITestNGMethod>,IConfigEavesdropper{}

public interface SuiteRunner extends ISuite, IInvokedMethodListener {}
```

> Instead, one could create a new test engine based on some internal TestNG code and we can help by providing needed classes in a dedicated library.

I am also suggesting the same, just that we dont have it as a dedicated library but just as an interface which our users can implement and a way to wire in these implementations.

**Comment by Travisnv** [comment]
Hi @krmahadevan  @juherr this is my question from google testng users group. I can give you guide a hand with PR. Rather copy code from main entry point, you guys can point me a draft plan for @krmahadevan's suggestion.

Let's say we want to create some thing like `@RunWith` annotation from JUnit5 for testng. How do you guys think about it?

**Comment by krmahadevan** [comment]
@Travisnv - I am in for being able to accept a customised testrunner/suiterunner implementation from a testng user. This has a lot of benefits (with the most ambitious of them being, the ability to build a distributed TestNG test execution environment). 

@juherr is mainly concerned that we don't end up in exposing a lot of things and his ask is that we perhaps need to come up with the capabilities of a TestNG engine and then have end users plugin their implementations into it.

If you have some ideas around how we can go about it, please feel free to create a skeleton draft PR and we can discuss more.

**Comment by juherr** [comment]
Frankly speaking, if you want to build a custom test engine, I recommend building over the junit5 platform and/or contributing to https://github.com/junit-team/testng-engine : you'll have more support there and it is designed with such extensions in mind.
IMO, TestNG should do less and be focused on its current killer features before adding new ones. 😉

**Comment by Travisnv** [comment]
@krmahadevan we can drive this request forward. From your exp how should we support a new annotation for suite runner (same with test runner)?

- On testng suite file
- On test class
- On test method

Other matter is multiple runners support? Order of runner?

**Comment by krmahadevan** [comment]
@Travisnv - Before we tread down that path, I think we should explore if we can use the [JUnit5 engine](https://github.com/junit-team/testng-engine) to achieve the same thing

**Comment by juherr** [comment]
Triage verdict: **still relevant**, but the scope is now narrower than the original issue.

Current TestNG already has a public `ITestRunnerFactory`, `TestNG.setTestRunnerFactoryClass(...)`, and CLI wiring for a custom test-runner factory. That covers much of the original `TestRunner` injection request. The suite side is still missing: `TestNG.createSuiteRunner(XmlSuite)` remains private and directly constructs `new SuiteRunner(...)`; there is no corresponding `ISuiteRunnerFactory` / suite-runner injection SPI.

Current tests exercise `ITestRunnerFactory` (including `TestRunnerTest` and the #2743 fixtures), but I found no current test or API for supplying a custom suite runner. I also found no #2632 / `GITHUB-2632` implementation, CHANGES entry, or open/merged PR that completes this request. The historical discussion considered using the JUnit Platform/TestNG engine instead, but no canonical replacement was established in this repository.

I did **not** execute a historical reproducer; this is an API enhancement and the missing suite-runner extension point is established directly from current code.

Next step: decide the public contract before implementation. In particular, define the minimal suite-runner capabilities that can be stable API, whether the factory should return `ISuite` rather than expose `SuiteRunner`, and how it composes with the existing `ITestRunnerFactory`. After that design decision, add API-level tests proving a custom suite-runner factory is selected without requiring users to subclass TestNG internals.
